code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -328,9 +373,233 @@ public long getServiceNetworkId(Service service) {
+                    LoadBalancer.class, lb.getId());
+            map = resourceMonitor.waitFor(map, new ResourcePredicate<LoadBalancerHostMap>() {
+                @Override
+                public boolean evaluate(LoadBalancerHostMap obj) {
+                    return obj != null && CommonStatesConstants.ACTIVE.equals(obj.getState());
+                }
+            });
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected List<Long> getLoadBalancerHostIds(Service service, int requestedScale, LoadBalancer lb) {
+        // TODO - once the swarm is in, swarm cluster is going to be sent instead of the hostIds
+        List<Host> availableHosts = objectManager.find(Host.class, HOST.ACCOUNT_ID, service.getAccountId(),
+                HOST.STATE, CommonStatesConstants.ACTIVE, HOST.REMOVED, null);
+        List<? extends LoadBalancerHostMap> existingHosts = mapDao.findNonRemoved(LoadBalancerHostMap.class,
+                LoadBalancer.class, lb.getId());
+        
+        List<Long> availableHostIds = (List<Long>) CollectionUtils.collect(availableHosts,
+                TransformerUtils.invokerTransformer("getId"));
+        List<Long> existingHostIds = (List<Long>) CollectionUtils.collect(existingHosts,
+                TransformerUtils.invokerTransformer("getId"));
+
+        requestedScale = requestedScale - existingHostIds.size();
+        availableHostIds.removeAll(existingHostIds);
+
+        if (availableHostIds.size() < requestedScale) {
+            throw new RuntimeException("Not enough hosts found  " + " to scale load balancer service "
+                    + service.getUuid());
+        }
+        Collections.shuffle(availableHostIds);
+        return availableHostIds;
+    }
+
+    @Override
+    public void deactivateLoadBalancerService(Service service) {
+        LoadBalancer lb = objectManager.findOne(LoadBalancer.class, LOAD_BALANCER.SERVICE_ID, service.getId(),
+                LOAD_BALANCER.REMOVED, null);
+        if (lb == null) {
+            return;
+        }
+        List<? extends LoadBalancerHostMap> maps = mapDao.findNonRemoved(LoadBalancerHostMap.class, LoadBalancer.class,
+                lb.getId());
+        for (LoadBalancerHostMap map : maps) {
+            lbService.removeHostFromLoadBalancer(lb, map.getHostId());
+        }
+    }
+
+    @Override
+    public void deactivateService(Service service) {
+        List<Instance> instances = objectManager.mappedChildren(service, Instance.class);
+        if (!instances.isEmpty()) {
+            stopServiceInstances(instances);
+        }
+    }
+
+    private void stopServiceInstances(List<Instance> instances) {
+        for (Instance instance : instances) {
+            objectProcessManager.scheduleProcessInstance(InstanceConstants.PROCESS_STOP, instance, null);
+        }
+    }
+
+    @Override
+    public void scaleDownService(Service service, int requestedScale) {
```

Comment:
@cjellick this will get fixed as a part of the separate PR. The original intention was to rely on the immutable service/environment names to be used for instances names. Then just today Vince/Darren suggested we let user to modify it. In this PR I've just moved this method from one class to another - it shouldn't affect regular service's behavior.